File: Flutter/packages/businessdash_sdk/lib/src/seo.dart
import 'dart:convert';

/// One tag for a document `<head>`.
class HeadTag {
  const HeadTag(this.tag, this.attributes, {this.children});

  final String tag;
  final List<MapEntry<String, String>> attributes;

  /// Text content — the title string, or a JSON-LD payload.
  final String? children;
}

/// The org's per-page SEO, turned into head tags.
///
/// The marketing page bundle carries a `seo` object for the page you asked
/// for: title, description, canonical, `noIndex`, Open Graph, Twitter card,
/// keywords, JSON-LD and hreflang. Reading four of those and forgetting the
/// rest is the normal failure, and the two people forget are `noIndex` and
/// `canonicalUrl` — exactly the two where being wrong is expensive and silent.
///
/// A Flutter app mostly has no `<head>`. This matters anyway for Flutter Web,
/// for a `WebView` rendering org content, and for populating a share sheet,
/// where `ogTitle` and `ogImageUrl` are the whole payload.
///
/// Rules, matching the TypeScript `seo-core`:
///
///  * Open Graph falls back to the page's own title/description. A missing
///    `og:title` renders a shared link as a bare URL.
///  * A RELATIVE canonical is dropped rather than emitted: crawlers resolve it
///    against whatever URL they fetched, so on a parameterised URL it points
///    somewhere nobody intended. A missing canonical is recoverable; a wrong
///    one consolidates ranking onto the wrong page.
///  * `robots` is always emitted, both ways. Absence means "index", so relying
///    on absence to express noindex is catastrophic.
List<HeadTag> seoHeadTags(Map<String, dynamic>? seo, {String? baseUrl}) {
  if (seo == null) return const [];

  final title = _text(seo['seoTitle']);
  final description = _text(seo['seoDescription']);
  final canonical = _absolute(_text(seo['canonicalUrl']), baseUrl);
  final noIndex = seo['noIndex'] == true;

  final ogTitle = _text(seo['ogTitle']) ?? title;
  final ogDescription = _text(seo['ogDescription']) ?? description;
  final ogImage = _absolute(_text(seo['ogImageUrl']), baseUrl);

  final tags = <HeadTag>[];
  if (title != null) tags.add(HeadTag('title', const [], children: title));
  if (description != null) tags.add(_meta('name', 'description', description));

  final keywords = seo['keywords'];
  if (keywords is List && keywords.isNotEmpty) {
    final joined = keywords.whereType<String>().join(', ');
    if (joined.isNotEmpty) tags.add(_meta('name', 'keywords', joined));
  }
  tags.add(_meta('name', 'robots', noIndex ? 'noindex, nofollow' : 'index, follow'));

  if (ogTitle != null) tags.add(_meta('property', 'og:title', ogTitle));
  if (ogDescription != null) tags.add(_meta('property', 'og:description', ogDescription));
  if (ogImage != null) tags.add(_meta('property', 'og:image', ogImage));
  if (canonical != null) tags.add(_meta('property', 'og:url', canonical));
  tags.add(_meta('property', 'og:type', 'website'));

  tags.add(_meta('name', 'twitter:card', _text(seo['twitterCard']) ?? 'summary_large_image'));
  if (ogTitle != null) tags.add(_meta('name', 'twitter:title', ogTitle));
  if (ogDescription != null) tags.add(_meta('name', 'twitter:description', ogDescription));
  if (ogImage != null) tags.add(_meta('name', 'twitter:image', ogImage));

  if (canonical != null) {
    tags.add(HeadTag('link', [
      const MapEntry('rel', 'canonical'),
      MapEntry('href', canonical),
    ]));
  }
  final hreflang = seo['hreflang'];
  if (hreflang is Map) {
    for (final entry in hreflang.entries) {
      final href = _absolute(_text(entry.value), baseUrl);
      if (href == null) continue;
      tags.add(HeadTag('link', [
        const MapEntry('rel', 'alternate'),
        MapEntry('hreflang', '${entry.key}'),
        MapEntry('href', href),
      ]));
    }
  }
  final nodes = seo['jsonldNodes'];
  if (nodes is List) {
    for (final node in nodes) {
      tags.add(HeadTag(
        'script',
        [const MapEntry('type', 'application/ld+json')],
        children: jsonEncode(node),
      ));
    }
  }
  return tags;
}

/// Ready-to-inject HTML, for Flutter Web or a WebView.
String renderSeoHead(Map<String, dynamic>? seo, {String? baseUrl}) {
  return seoHeadTags(seo, baseUrl: baseUrl).map((tag) {
    final attrs =
        tag.attributes.map((a) => '${a.key}="${_escape(a.value)}"').join(' ');
    switch (tag.tag) {
      case 'title':
        return '<title>${_escape(tag.children ?? '')}</title>';
      case 'script':
        // `<` is what closes the script early; escaping the whole payload
        // would corrupt the JSON instead.
        final payload = (tag.children ?? '').replaceAll('<', r'\u003c');
        return '<script $attrs>$payload</script>';
      default:
        return '<${tag.tag} $attrs>';
    }
  }).join('\n');
}

HeadTag _meta(String key, String name, String content) =>
    HeadTag('meta', [MapEntry(key, name), MapEntry('content', content)]);

String? _text(dynamic value) {
  if (value is! String) return null;
  final trimmed = value.trim();
  return trimmed.isEmpty ? null : trimmed;
}

String? _absolute(String? url, String? baseUrl) {
  if (url == null) return null;
  if (RegExp(r'^([a-z][a-z0-9+.-]*:|//)', caseSensitive: false).hasMatch(url)) {
    return url;
  }
  // A relative canonical is worse than none — see the note above.
  if (baseUrl == null || baseUrl.trim().isEmpty) return null;
  final base = baseUrl.replaceAll(RegExp(r'/+$'), '');
  return url.startsWith('/') ? '$base$url' : '$base/$url';
}

String _escape(String value) => value
    .replaceAll('&', '&amp;')
    .replaceAll('"', '&quot;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;');

File: Kotlin/biab/src/main/kotlin/app/biab/Seo.kt
package app.biab

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull

/** One tag for a document `<head>`. */
public data class HeadTag(
    val tag: String,
    val attributes: List<Pair<String, String>>,
    /** Text content — the title string, or a JSON-LD payload. */
    val children: String? = null,
)

/**
 * The org's per-page SEO, turned into head tags.
 *
 * The marketing page bundle carries a `seo` object for the page you asked for:
 * title, description, canonical, `noIndex`, Open Graph, Twitter card,
 * keywords, JSON-LD and hreflang. Reading four of those and forgetting the
 * rest is the normal failure, and the two people forget are `noIndex` and
 * `canonicalUrl` — exactly the two where being wrong is expensive and silent.
 *
 * An Android app mostly has no `<head>`. This matters anyway for a WebView
 * rendering org content, and for populating a share intent, where `ogTitle`
 * and `ogImageUrl` are the whole payload.
 *
 * Rules, matching the TypeScript `seo-core`:
 *
 *  - Open Graph falls back to the page's own title/description. A missing
 *    `og:title` renders a shared link as a bare URL.
 *  - A RELATIVE canonical is dropped rather than emitted: crawlers resolve it
 *    against whatever URL they fetched, so on a parameterised URL it points
 *    somewhere nobody intended. A missing canonical is recoverable; a wrong one
 *    consolidates ranking onto the wrong page.
 *  - `robots` is always emitted, both ways. Absence means "index", so relying
 *    on absence to express noindex is catastrophic.
 */
public object Seo {
    public fun headTags(seo: JsonObject?, baseUrl: String? = null): List<HeadTag> {
        if (seo == null) return emptyList()

        val title = text(seo["seoTitle"])
        val description = text(seo["seoDescription"])
        val canonical = absolute(text(seo["canonicalUrl"]), baseUrl)
        val noIndex = (seo["noIndex"] as? JsonPrimitive)?.booleanOrNull == true

        val ogTitle = text(seo["ogTitle"]) ?: title
        val ogDescription = text(seo["ogDescription"]) ?: description
        val ogImage = absolute(text(seo["ogImageUrl"]), baseUrl)

        val tags = mutableListOf<HeadTag>()
        title?.let { tags += HeadTag("title", emptyList(), it) }
        description?.let { tags += meta("name", "description", it) }
        (seo["keywords"] as? JsonArray)
            ?.mapNotNull { text(it) }
            ?.takeIf { it.isNotEmpty() }
            ?.let { tags += meta("name", "keywords", it.joinToString(", ")) }
        tags += meta("name", "robots", if (noIndex) "noindex, nofollow" else "index, follow")

        ogTitle?.let { tags += meta("property", "og:title", it) }
        ogDescription?.let { tags += meta("property", "og:description", it) }
        ogImage?.let { tags += meta("property", "og:image", it) }
        canonical?.let { tags += meta("property", "og:url", it) }
        tags += meta("property", "og:type", "website")

        tags += meta("name", "twitter:card", text(seo["twitterCard"]) ?: "summary_large_image")
        ogTitle?.let { tags += meta("name", "twitter:title", it) }
        ogDescription?.let { tags += meta("name", "twitter:description", it) }
        ogImage?.let { tags += meta("name", "twitter:image", it) }

        canonical?.let {
            tags += HeadTag("link", listOf("rel" to "canonical", "href" to it))
        }
        (seo["hreflang"] as? JsonObject)?.forEach { (lang, href) ->
            absolute(text(href), baseUrl)?.let {
                tags += HeadTag("link", listOf("rel" to "alternate", "hreflang" to lang, "href" to it))
            }
        }
        (seo["jsonldNodes"] as? JsonArray)?.forEach { node ->
            tags += HeadTag("script", listOf("type" to "application/ld+json"), node.toString())
        }
        return tags
    }

    /** Ready-to-inject HTML, for a WebView or a server-rendered template. */
    public fun render(seo: JsonObject?, baseUrl: String? = null): String =
        headTags(seo, baseUrl).joinToString("\n") { tag ->
            val attrs = tag.attributes.joinToString(" ") { (k, v) -> "$k=\"${escape(v)}\"" }
            when (tag.tag) {
                "title" -> "<title>${escape(tag.children.orEmpty())}</title>"
                // `<` is what closes the script early; escaping the whole
                // payload would corrupt the JSON instead.
                "script" -> "<script $attrs>${tag.children.orEmpty().replace("<", "\\u003c")}</script>"
                else -> "<${tag.tag} $attrs>"
            }
        }

    private fun meta(key: String, name: String, content: String) =
        HeadTag("meta", listOf(key to name, "content" to content))

    private fun text(value: JsonElement?): String? =
        (value as? JsonPrimitive)?.contentOrNull?.trim()?.takeIf { it.isNotEmpty() }

    private fun absolute(url: String?, baseUrl: String?): String? {
        if (url == null) return null
        if (Regex("^([a-z][a-z0-9+.-]*:|//)", RegexOption.IGNORE_CASE).containsMatchIn(url)) return url
        // A relative canonical is worse than none — see the note above.
        if (baseUrl.isNullOrBlank()) return null
        val base = baseUrl.trimEnd('/')
        return if (url.startsWith("/")) base + url else "$base/$url"
    }

    private fun escape(value: String): String =
        value.replace("&", "&amp;").replace("\"", "&quot;")
            .replace("<", "&lt;").replace(">", "&gt;")
}
